WordPress/gutenberg
# Add block spacing gap config to theme.json and add support for this CSS variable to the "flow/default" layout.

Folks have asked for some time now for a consistent way to tweak the margin/gap between blocks. This PR tries to solve that.

**Decisions and choices made on this PR**

 - The "block gap" is a generic style that can apply to any container block no matter of its layout type (right now we have flow and flex layouts but more will come)
 - Each layout type can "consume" the gap configuration and CSS variable and apply it to its children properly. In the current PR, the "flow" (or default) layout type uses the gap config to space the blocks vertically. While not implemented yet the "flex horizontal" layout could use the gap to space the blocks horizontally.
 - No UI/block support has been added for now, but we could imagine the gap being added to the inspector controls of container blocks (regardless of their layout type) to tweak its value contextually.
 - We'll provide a default style for this property for themes with theme.json file. This is most likely going to impact themes that are trying to switch to theme.json. I think the impact is not going to be very big but it's worth mentioning. 

**Notes**

For block templates a global CSS style applied to the root of the block templates is added by default to the theme.json generated stylesheet. Unfortunately for classic templates, we can't generate that styles automatically in the frontend because the "wrapper" for post-content doesn't have a consistent className (like `.entry-content` or something). This is also the exact same reason we ask these theme-authors to add "layout" (wide/full alignments) styles on their own to the frontend. We could potentially try a solution for this separately by relying on a convention.

**Testing instructions**

 - tweak the `"spacing": { "gap": "some value" }` in your theme.json file
 - Ensure that it impacts editor and frontend properly 
 - Check that there's no specific block styles overriding this gap.
